### PR TITLE
drivers: sensors: add MC3419 accel sensor support

### PR DESCRIPTION
**Pull request by karthi012**

### CODE REVIEW

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+	gpio_pin_interrupt_configure_dt(&cfg->int_gpio, GPIO_INT_DISABLE);
+#if defined CONFIG_MC3419_TRIGGER_OWN_THREAD
+	k_sem_give(&data->trig_sem);
+#elif defined CONFIG_MC3419_TRIGGER_GLOBAL_THREAD
+	k_work_submit(&data->work);
+#endif
+}
+
+static void mc3419_drdy_interrupt_handler(const struct device *dev)
```
Comment (ssekar15) [inline drivers/sensor/mc3419/mc3419_trigger.c:33]: use of array avoids these wrapper

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+#if defined CONFIG_MC3419_MOTION
+	if (int_source & MC3419_ANY_MOTION_MASK) {
+		mc3419_motion_interrupt_handler(dev);
+	}
+#endif
+
+exit:
+	ret = i2c_reg_write_byte_dt(&cfg->i2c, MC3419_REG_INT_STATUS, MC3419_INT_CLEAR);
+	if (ret < 0) {
+		goto exit;
```
Comment (ssekar15) [inline drivers/sensor/mc3419/mc3419_trigger.c:76]: code in loop.

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+	}
+
+	gpio_pin_interrupt_configure_dt(&cfg->int_gpio, GPIO_INT_EDGE_FALLING);
```
Comment (ssekar15) [inline drivers/sensor/mc3419/mc3419_trigger.c:79]: why should configure on every interrupt?

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+	int ret = 0;
+	uint8_t buf = 0;
+	const struct mc3419_config *cfg = dev->config;
+	struct mc3419_driver_data *data = dev->data;
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	ret = mc3419_set_op_mode(cfg, MC3419_MODE_STANDBY);
+	if (ret < 0) {
+		return ret;
```
Comment (ssekar15) [inline drivers/sensor/mc3419/mc3419_trigger.c:115]: release acquired lock

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+				sensor_trigger_handler_t handler)
+{
+	int ret = 0;
+	uint8_t buf = 0;
+	const struct mc3419_config *cfg = dev->config;
+	struct mc3419_driver_data *data = dev->data;
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	ret = mc3419_set_op_mode(cfg, MC3419_MODE_STANDBY);
```
Comment (ssekar15) [inline drivers/sensor/mc3419/mc3419_trigger.c:113]: better to introduce a wrapper with semaphore and op_mode  in  mc3419.c

File: drivers/sensor/mc3419/mc3419.c [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+	k_sem_take(&data->sem, K_FOREVER);
+	ret = i2c_burst_read_dt(&cfg->i2c, MC3419_REG_XOUT_L,
+				(uint8_t *)data->samples,
+				MC3419_SAMPLE_READ_SIZE);
+	k_sem_give(&data->sem);
+	return ret;
+}
+
+static void mc3419_data_convert(double sensitivity, int16_t *raw_data,
```
Comment (parthitce) [inline drivers/sensor/mc3419/mc3419.c:57]: `mc3419_to_sensor_value` may be?

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+#if defined(CONFIG_MC3419_TRIGGER_OWN_THREAD)
+	k_sem_init(&data->trig_sem, 0, 1);
+	k_thread_create(&data->thread, data->thread_stack,
+			CONFIG_MC3419_THREAD_STACK_SIZE,
+			(k_thread_entry_t)mc3419_thread, data, NULL,
+			NULL, K_PRIO_COOP(CONFIG_MC3419_THREAD_PRIORITY), 0,
+			K_NO_WAIT);
+#else
+	data->work.handler = mc3419_work_cb;
```
Comment (parthitce) [inline drivers/sensor/mc3419/mc3419_trigger.c:158]:
```
void k_work_init(struct k_work *work,
                  k_work_handler_t handler)
{
        __ASSERT_NO_MSG(work != NULL);
        __ASSERT_NO_MSG(handler != NULL);

        *work = (struct k_work)Z_WORK_INITIALIZER(handler);

        SYS_PORT_TRACING_OBJ_INIT(k_work, work);
}
```
use `k_work_init`

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+#endif
+	gpio_init_callback(&data->gpio_cb, mc3419_gpio_callback,
+			   BIT(cfg->int_gpio.pin));
+	ret = gpio_add_callback(cfg->int_gpio.port, &data->gpio_cb);
+	if (ret < 0) {
+		LOG_ERR("Failed to set int callback");
+		return ret;
+	}
+
+#if defined(CONFIG_MC3419_DRDY_INT2) || defined(CONFIG_MC3419_MOTION_INT2)
```
Comment (parthitce) [inline drivers/sensor/mc3419/mc3419_trigger.c:168]: How does it default to INT1?

File: dts/bindings/sensor/memsic,mc3419.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+description: MC3419 3-axis accel sensor
+
+compatible: "memsic,mc3419"
+
+include: [sensor-device.yaml, i2c-device.yaml]
+
+properties:
+  int-gpios:
+    type: phandle-array
+    required: true
```
Comment (parthitce) [inline dts/bindings/sensor/memsic,mc3419.yaml:13]: interrupt must be optional

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+	const struct mc3419_config *cfg = dev->config;
+	struct mc3419_driver_data *data = dev->data;
+
+	if (trig->type & SENSOR_TRIG_DATA_READY) {
+		data->handler[MC3419_TRIG_DATA_READY] = handler;
+		data->trigger[MC3419_TRIG_DATA_READY] = trig;
+		buf |= MC3419_DATA_READY_MASK;
+	}
+
+	if (trig->type & SENSOR_TRIG_MOTION) {
```
Comment (parthitce) [inline drivers/sensor/mc3419/mc3419_trigger.c:98]: `else if` ?

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+		uint8_t int_mask = MC3419_ANY_MOTION_MASK;
+
+		buf |= MC3419_ANY_MOTION_MASK;
+		data->handler[MC3419_TRIG_ANY_MOTION] = handler;
+		data->trigger[MC3419_TRIG_ANY_MOTION] = trig;
+
+		ret = i2c_reg_update_byte_dt(&cfg->i2c, MC3419_REG_MOTION_CTRL,
```
Comment (parthitce) [inline drivers/sensor/mc3419/mc3419_trigger.c:105]: doesn't need to be in MC3419_MODE_STANDBY before configure?
`ret = mc3419_set_op_mode(dev->config, MC3419_MODE_STANDBY);`

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+			return ret;
+		}
+	}
+	if (!(trig->type & SENSOR_TRIG_DATA_READY) &&
+	    !(trig->type & SENSOR_TRIG_MOTION)) {
+		LOG_ERR("Unsupported sensor trigger");
+		return -ENOTSUP;
+	}
+
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, MC3419_REG_INT_CTRL,
```
Comment (parthitce) [inline drivers/sensor/mc3419/mc3419_trigger.c:118]: ditto `MC3419_MODE_STANDBY` ?

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+				     MC3419_DATA_READY_MASK | MC3419_ANY_MOTION_MASK,
+				     buf);
+	if (ret < 0) {
+		LOG_ERR("Failed to configure interrupt (%d)", ret);
+		return ret;
+	}
+
+	gpio_pin_interrupt_configure_dt(&cfg->int_gpio, GPIO_INT_EDGE_FALLING);
```
Comment (parthitce) [inline drivers/sensor/mc3419/mc3419_trigger.c:126]: Any specific reason for falling edge? Not all the SoC's support edge triggered interrupt in GPIO

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+		data->trigger[MC3419_TRIG_ANY_MOTION] = trig;
+
+		ret = i2c_reg_update_byte_dt(&cfg->i2c, MC3419_REG_MOTION_CTRL,
+					     int_mask, handler ? int_mask : 0);
+		if (ret < 0) {
+			LOG_ERR("Failed to configure motion interrupt (%d)", ret);
+			return ret;
+		}
+	}
+	if (!(trig->type & SENSOR_TRIG_DATA_READY) &&
```
Comment (parthitce) [inline drivers/sensor/mc3419/mc3419_trigger.c:112]: Move to entry of the function

File: drivers/sensor/mc3419/mc3419_trigger.c [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	if (int_source & MC3419_DATA_READY_MASK) {
+		if (data->handler[MC3419_TRIG_DATA_READY]) {
+			data->handler[MC3419_TRIG_DATA_READY](dev,
+				data->trigger[MC3419_TRIG_DATA_READY]);
+		}
+	}
+
+	if (int_source & MC3419_ANY_MOTION_MASK) {
```
Comment (parthitce) [inline drivers/sensor/mc3419/mc3419_trigger.c:51]: `else if`  or both interrupt can occur at the same time?

**Review by parthitce — COMMENTED** [review]
Sensor sample is missing.

File: drivers/sensor/mc3419/mc3419_trigger.c (hunk 15 above)
Comment (ssekar15) [reply drivers/sensor/mc3419/mc3419_trigger.c:51]: decision is based on status register and will be cleared on read so good to check both imo

**Comment by ssekar15** [comment]
looks fine for me

**Pull request closed without merging**